File: skypydb-js/src/database/mixins/vector/collections/collection_audit.ts
import type Database from "better-sqlite3";
import { InputValidator } from "../../../../security/validation";

type ItemRow = {
  id: string;
  document: string | null;
  embedding: number[];
  metadata: Record<string, unknown> | null;
  created_at?: string;
};

export abstract class CollectionAuditMixin {
  protected abstract conn: Database.Database;

  collection_exists(name: string): boolean {
    const validated = InputValidator.validate_table_name(name);
    const row = this.conn
      .prepare(
        "SELECT name FROM sqlite_master WHERE type='table' AND name = ?"
      )
      .get(`vec_${validated}`);
    return row !== undefined;
  }

  protected _ensure_collections_table(): void {
    this.conn
      .prepare(
        `
        CREATE TABLE IF NOT EXISTS _vector_collections (
          name TEXT PRIMARY KEY,
          metadata TEXT,
          created_at TEXT NOT NULL
        )
        `
      )
      .run();
  }

  protected _matches_filters(
    item: ItemRow,
    where?: Record<string, unknown>,
    where_document?: Record<string, string>
  ): boolean {
    if (where) {
      const metadata = item.metadata ?? {};
      for (const [key, value] of Object.entries(where)) {
        if (key.startsWith("$")) {
          if (key === "$and") {
            const conditions = (value as Array<Record<string, unknown>>) ?? [];
            if (!conditions.every((condition) => this._matches_filters(item, condition, undefined))) {
              return false;
            }
          } else if (key === "$or") {
            const conditions = (value as Array<Record<string, unknown>>) ?? [];
            if (!conditions.some((condition) => this._matches_filters(item, condition, undefined))) {
              return false;
            }
          }
          continue;
        }

        if (value && typeof value === "object" && !Array.isArray(value)) {
          const meta_value = metadata[key];
          for (const [operator, operator_value] of Object.entries(value)) {
            if (operator === "$eq" && meta_value !== operator_value) {
              return false;
            }
            if (operator === "$ne" && meta_value === operator_value) {
              return false;
            }
            if (operator === "$gt" && !(meta_value !== undefined && (meta_value as number) > (operator_value as number))) {
              return false;
            }
            if (operator === "$gte" && !(meta_value !== undefined && (meta_value as number) >= (operator_value as number))) {
              return false;
            }
            if (operator === "$lt" && !(meta_value !== undefined && (meta_value as number) < (operator_value as number))) {
              return false;
            }
            if (operator === "$lte" && !(meta_value !== undefined && (meta_value as number) <= (operator_value as number))) {
              return false;
            }
            if (operator === "$in" && Array.isArray(operator_value) && !operator_value.includes(meta_value)) {
              return false;
            }
            if (operator === "$nin" && Array.isArray(operator_value) && operator_value.includes(meta_value)) {
              return false;
            }
          }
        } else if (metadata[key] !== value) {
          return false;
        }
      }
    }

    if (where_document) {
      const document = item.document ?? "";
      for (const [operator, value] of Object.entries(where_document)) {
        if (operator === "$contains" && !document.includes(value)) {
          return false;
        }
        if (operator === "$not_contains" && document.includes(value)) {
          return false;
        }
      }
    }

    return true;
  }
}

File: skypydb/database/operations/vector/collections/collection_audit.py
"""Shared collection metadata helpers for vector storage."""

from typing import Any, Dict, Optional

from skypydb.security.validation import InputValidator


class CollectionAuditMixin:
    """Provide common collection integrity and filter helpers."""

    def collection_exists(self, name: str) -> bool:
        """Return whether the collection table exists."""

        name = InputValidator.validate_table_name(name)
        cursor = self.conn.cursor()
        cursor.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?",
            (f"vec_{name}",),
        )
        return cursor.fetchone() is not None

    def _ensure_collections_table(self) -> None:
        """Ensure the metadata registry table for collections exists."""

        cursor = self.conn.cursor()
        cursor.execute(
            """
            CREATE TABLE IF NOT EXISTS _vector_collections (
                name TEXT PRIMARY KEY,
                metadata TEXT,
                created_at TEXT NOT NULL
            )
            """
        )
        self.conn.commit()

    def _matches_filters(
        self,
        item: Dict[str, Any],
        where: Optional[Dict[str, Any]] = None,
        where_document: Optional[Dict[str, str]] = None,
    ) -> bool:
        """Evaluate metadata and document predicates against an item."""

        if where is not None:
            metadata = item.get("metadata") or {}

            # Metadata operators are evaluated recursively to support nested `$and`/`$or` groups.
            for key, value in where.items():
                if key.startswith("$"):
                    if key == "$and":
                        if not all(self._matches_filters(item, cond, None) for cond in value):
                            return False
                    elif key == "$or":
                        if not any(self._matches_filters(item, cond, None) for cond in value):
                            return False
                    continue

                if isinstance(value, dict):
                    meta_value = metadata.get(key)
                    for operator, op_value in value.items():
                        if operator == "$eq" and meta_value != op_value:
                            return False
                        elif operator == "$ne" and meta_value == op_value:
                            return False
                        elif operator == "$gt" and not (
                            meta_value is not None and meta_value > op_value
                        ):
                            return False
                        elif operator == "$gte" and not (
                            meta_value is not None and meta_value >= op_value
                        ):
                            return False
                        elif operator == "$lt" and not (
                            meta_value is not None and meta_value < op_value
                        ):
                            return False
                        elif operator == "$lte" and not (
                            meta_value is not None and meta_value <= op_value
                        ):
                            return False
                        elif operator == "$in" and meta_value not in op_value:
                            return False
                        elif operator == "$nin" and meta_value in op_value:
                            return False
                elif metadata.get(key) != value:
                    return False

        if where_document is not None:
            document = item.get("document") or ""
            for operator, value in where_document.items():
                if operator == "$contains" and value not in document:
                    return False
                if operator == "$not_contains" and value in document:
                    return False

        return True
